libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3158,6 +3158,12 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 
 			annotation.isPure = isPure;
 		}
+	if (
+		VariableDeclaration const* varDecl = dynamic_cast<decltype(varDecl)>(annotation.referencedDeclaration);
+		!annotation.isPure.set() &&
```

Comment:
This looks like it depends on whether it is a forward-reference or not - I'm not sure we really want that.

Reply:
How does it work for file-level constants?

Reply:
> This looks like it depends on whether it is a forward-reference or not - I'm not sure we really want that.

So we basically need a separate pass that walks the AST, keeps track of dependencies and can go back to them? Or maybe just does it recursively?

> How does it work for file-level constants?

Yeah, I think we need test cases covering that. Also a mix between file-level and contract constants.

Reply:
These are good concerns, but fortunately I think we can bypass them. TL;DR: we should probably use `varDecl->isConstant()` here.

At the type checker level, when we see a member reference, we should simply respect its declaration type. In this case, use `varDecl->isConstant()`. Afaik this is what most of the type checkers do, and what they should do (else you need type inference).

This way, we don't need to worry about annotation of the actual `varDecl->value()`, i.e., if it's not pure, type checker will complain when visiting that `varDecl`.

If we have a cycle, [ConstStateVarCircularReferenceChecker](https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/PostTypeChecker.cpp#L148-L149) will take care of it.

I included 3 tests which should cover all the file/contract level constant interactions.

Reply:
Great! This looks like a much better approach.